README.md
# sized_matrix

[Crate](https://crates.io/crates/sized_matrix)

[Documentation](https://docs.rs/sized_matrix)

[Repository](https://github.com/LukeMiles49/sized-matrix-rs)

[Changelog](https://github.com/LukeMiles49/sized-matrix-rs/blob/master/CHANGELOG.md)

Sized matrices using const generics for better type checking and performance.

```rust
use sized_matrix::{Matrix, Vector};

let a: Matrix<i32, 3, 4> = Matrix::rows([
	[ 1,  2,  3,  4],
	[ 5,  6,  7,  8],
	[ 9, 10, 11, 12],
]);

let b: Matrix<i32, 4, 2> = Matrix::rows([
	[ 0,  1],
	[ 1,  2],
	[ 3,  5],
	[ 8, 13],
]);

let c: Matrix<i32, 3, 2> = a * b;

assert_eq!(c, Matrix::rows([
	[ 43,  72],
	[ 91, 156],
	[139, 240],
]));

let d: Vector<i32, 2> = Matrix::vector([-1, 1]);

let e: Vector<i32, 3> = c * d;

assert_eq!(e, Matrix::vector([
	 29,
	 65,
	101,
]));
```

To use this, add it as a dependency to your Cargo.toml:
```toml
[dependencies]
sized_matrix = "^0.3.0"
```


CHANGELOG.md
# Changelog

This project follows [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.3.0](https://crates.io/crates/sized_matrix/0.3.0) - 2022-06-27

### Updated:
* Updated to [higher_order_functions](https://crates.io/crates/higher_order_functions) v0.2.0
* Removed unnecessary unstable features

## [v0.2.5](https://crates.io/crates/sized_matrix/0.2.5) - 2021-04-05

### Fixed:
* Typo in `Vector` `IndexMut` causing index out of bounds error

## [v0.2.4](https://crates.io/crates/sized_matrix/0.2.4) - 2021-04-04

### Fixed:
* Unstable function name changed by <https://github.com/rust-lang/rust/issues/63567>

## [v0.2.3](https://crates.io/crates/sized_matrix/0.2.3) - 2020-11-12

### Added:
* Added left-division for `Matrix` (`rhs^-1 * self` instead of `self * rhs^-1`)

## [v0.2.2](https://crates.io/crates/sized_matrix/0.2.2) - 2020-09-08

### Added:
* Implemented `Zip` and `Section` from [higher_order_functions](https://crates.io/crates/higher_order_functions) for `Vector` and `Matrix`

## [v0.2.1](https://crates.io/crates/sized_matrix/0.2.1) - 2020-09-06

### Added:
* Dot product for vectors

## [v0.2.0](https://crates.io/crates/sized_matrix/0.2.0) - 2020-09-04

### Changed:
* Moved from [init_trait](https://crates.io/crates/init_trait) to [higher_order_functions](https://crates.io/crates/higher_order_functions)

### Added:
* Implemented [Map](https://docs.rs/higher_order_functions/0.1.0/higher_order_functions/trait.Map.html) for matrices

## [v0.1.0](https://crates.io/crates/sized_matrix/0.1.0) - 2020-08-23

### Added:
* Initial implementation


tests/matrix.rs
use sized_matrix::{Matrix, Transpose};
use higher_order_functions::{Init, Map, Zip, Section};

use core::hash::{Hash, Hasher};
use std::collections::hash_map::DefaultHasher;
use std::panic::{catch_unwind, UnwindSafe};

use num_traits::{
	One, Zero,
	Inv, Pow,
	MulAdd, MulAddAssign,
};

fn assert_panics<R, F: FnOnce() -> R + UnwindSafe>(f: F) {
	assert!(catch_unwind(f).is_err());
}

#[test]
fn init_matrix() {
	let matrix = Matrix::<_, 3, 5>::init(|[row, col]| (row, col));
	
	for row in 0..3 {
		for col in 0..5 {
			assert_eq!(matrix[[row, col]], (row, col));
		}
	}
}

#[test]
fn matrix_rows() {
	let matrix = Matrix::rows([
		[(0, 0), (0, 1)],
		[(1, 0), (1, 1)],
		[(2, 0), (2, 1)],
	]);
	
	assert_eq!(matrix, Matrix::<_, 3, 2>::init(|[row, col]| (row, col)));
}

#[test]
fn matrix_cols() {
	let matrix = Matrix::cols([
		[(0, 0), (1, 0), (2, 0)],
		[(0, 1), (1, 1), (2, 1)],
	]);
	
	assert_eq!(matrix, Matrix::<_, 3, 2>::init(|[row, col]| (row, col)));
}

#[test]
fn matrix_from_vectors() {
	let matrix = Matrix::from_vectors([
		Matrix::vector([(0, 0), (1, 0), (2, 0)]),
		Matrix::vector([(0, 1), (1, 1), (2, 1)]),
	]);
	
	assert_eq!(matrix, Matrix::<_, 3, 2>::init(|[row, col]| (row, col)));
}

#[test]
fn matrix_map_double() {
	let matrix = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	assert_eq!(matrix.map(|x| x * 2), Matrix::rows([[2, 8, 10], [4, 6, 12]]));
}

#[test]
fn matrix_map_cast() {
	let matrix = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	assert_eq!(matrix.map(f64::from), Matrix::rows([[1.0, 4.0, 5.0], [2.0, 3.0, 6.0]]));
}

#[test]
fn empty_matrix_section() {
	let m: Matrix<(), 0, 0> = Matrix::rows([]);
	
	let mat: Matrix<_, 0, 0> = m.section([0, 0]);
	
	assert_eq!(mat, Matrix::rows([]));
}

#[test]
fn matrix_empty_section() {
	let m = Matrix::<_, 3, 5>::init(|[row, col]| (row, col));
	
	assert_eq!(m.section([1, 2]), Matrix::<_, 0, 0>::rows([]));
	assert_eq!(m.section([1, 2]), Matrix::<_, 2, 0>::cols([]));
	assert_eq!(m.section([1, 2]), Matrix::<_, 0, 2>::rows([]));
}

#[test]
fn matrix_empty_out_of_bounds() {
	let m = Matrix::<_, 3, 5>::init(|[row, col]| (row, col));
	
	assert_panics(|| m.section([4, 2]) as Matrix<_, 0, 0>);
	assert_panics(|| m.section([1, 6]) as Matrix<_, 0, 0>);
}

#[test]
fn matrix_middle_section() {
	let m = Matrix::<_, 3, 5>::init(|[row, col]| (row, col));
	
	assert_eq!(m.section([1, 1]), Matrix::rows([[(1, 1), (1, 2), (1, 3)]]));
}

#[test]
fn matrix_corner_sections() {
	let m = Matrix::<_, 3, 5>::init(|[row, col]| (row, col));
	
	assert_eq!(m.section([0, 0]), Matrix::rows([[(0, 0), (0, 1)], [(1, 0), (1, 1)]]));
	assert_eq!(m.section([0, 3]), Matrix::rows([[(0, 3), (0, 4)], [(1, 3), (1, 4)]]));
	assert_eq!(m.section([1, 3]), Matrix::rows([[(1, 3), (1, 4)], [(2, 3), (2, 4)]]));
	assert_eq!(m.section([1, 0]), Matrix::rows([[(1, 0), (1, 1)], [(2, 0), (2, 1)]]));
}

#[test]
fn matrix_side_sections() {
	let m = Matrix::<_, 3, 5>::init(|[row, col]| (row, col));
	
	assert_eq!(m.section([0, 0]), Matrix::rows([[(0, 0), (0, 1), (0, 2), (0, 3), (0, 4)]]));
	assert_eq!(m.section([0, 4]), Matrix::rows([[(0, 4)], [(1, 4)], [(2, 4)]]));
	assert_eq!(m.section([2, 0]), Matrix::rows([[(2, 0), (2, 1), (2, 2), (2, 3), (2, 4)]]));
	assert_eq!(m.section([0, 0]), Matrix::rows([[(0, 0)], [(1, 0)], [(2, 0)]]));
}

#[test]
fn matrix_full_section() {
	let m = Matrix::<_, 3, 5>::init(|[row, col]| (row, col));
	
	assert_eq!(m.section([0, 0]), Matrix::<_, 3, 5>::init(|[row, col]| (row, col)));
}

#[test]
fn empty_matrix_zip() {
	let a: Matrix<(), 0, 0> = Matrix::rows([]);
	let b: Matrix<(), 0, 0> = Matrix::rows([]);
	
	let mat = a.zip_with(b, |_: (), _: ()| panic!("Shouldn't call zip function"));
	
	assert_eq!(mat, Matrix::rows([]));
}

#[test]
fn singleton_matrix_zip() {
	let a = Matrix::rows([[123]]);
	let b = Matrix::rows([[456]]);
	let mut called = false;
	
	let arr = a.zip_with(b, |a, b| {
		assert_eq!(a, 123);
		assert_eq!(b, 456);
		if called { panic!("Should only call zip function once"); }
		else { called = true; }
		321
	});
	
	assert!(called);
	assert_eq!(arr, Matrix::rows([[321]]));
}

#[test]
fn matrix_zip() {
	let a = Matrix::<_, 3, 5>::init(|[row, col]| (row, col, "a"));
	let b = Matrix::<_, 3, 5>::init(|[row, col]| (row, col, "b"));
	
	let arr = a.zip_with(b, |ax, bx| (ax, bx));
	
	assert_eq!(arr, Matrix::<_, 3, 5>::init(|[row, col]| ((row, col, "a"), (row, col, "b"))));
}

#[test]
fn matrix_zero() {
	let matrix = Matrix::<usize, 3, 5>::zero();
	
	assert_eq!(matrix, Matrix::<_, 3, 5>::init(|_| 0));
}

#[test]
fn matrix_one() {
	let matrix = Matrix::<usize, 3, 3>::one();
	
	assert_eq!(matrix, Matrix::<_, 3, 3>::init(|[row, col]| if row == col { 1 } else { 0 }));
}

#[test]
fn matrix_eq() {
	assert_eq!(Matrix::rows([[1, 2], [3, 4]]), Matrix::rows([[1, 2], [3, 4]]));
	assert_ne!(Matrix::rows([[1, 2], [3, 4]]), Matrix::rows([[1, 2], [3, 0]]));
	assert_ne!(Matrix::rows([[1, 2], [3, 4]]), Matrix::rows([[1, 2], [0, 4]]));
	assert_ne!(Matrix::rows([[1, 2], [3, 4]]), Matrix::rows([[1, 0], [3, 4]]));
	assert_ne!(Matrix::rows([[1, 2], [3, 4]]), Matrix::rows([[0, 2], [3, 4]]));
	assert_eq!(Matrix::<(), 0, 0>::rows([]), Matrix::<(), 0, 0>::rows([]));
}

#[test]
fn matrix_hash() {
	fn hash<T: Hash>(t: T) -> u64 {
		let mut hasher = DefaultHasher::new();
		t.hash(&mut hasher);
		hasher.finish()
	}
	
	// This isn't technically guaranteed to succeed, but I can't think of a better way to test it
	assert_eq!(hash(Matrix::rows([[1, 2], [3, 4]])), hash(Matrix::rows([[1, 2], [3, 4]])));
	assert_ne!(hash(Matrix::rows([[1, 2], [3, 4]])), hash(Matrix::rows([[1, 2], [3, 0]])));
	assert_ne!(hash(Matrix::rows([[1, 2], [3, 4]])), hash(Matrix::rows([[1, 2], [0, 4]])));
	assert_ne!(hash(Matrix::rows([[1, 2], [3, 4]])), hash(Matrix::rows([[1, 0], [3, 4]])));
	assert_ne!(hash(Matrix::rows([[1, 2], [3, 4]])), hash(Matrix::rows([[0, 2], [3, 4]])));
	assert_eq!(hash(Matrix::<(), 0, 0>::rows([])), hash(Matrix::<(), 0, 0>::rows([])));
}

#[test]
fn matrix_debug() {
	assert_eq!(
		format!("{:?}", Matrix::<_, 3, 2>::init(|[row, col]| (row, col))),
		"Matrix([[(0, 0), (0, 1)], [(1, 0), (1, 1)], [(2, 0), (2, 1)]])",
	);
	
	assert_eq!(format!("{:?}", Matrix::<(), 0, 0>::rows([])), "Matrix([])");
}

#[test]
fn matrix_default() {
	let matrix = Matrix::<usize, 3, 5>::default();
	
	assert_eq!(matrix, Matrix::<_, 3, 5>::init(|_| 0));
}

#[test]
fn matrix_transpose() {
	let matrix = Matrix::<_, 3, 5>::init(|[row, col]| (row, col)).transpose();
	
	assert_eq!(matrix, Matrix::<_, 5, 3>::init(|[row, col]| (col, row)));
}

#[test]
fn matrix_mul_scalar() {
	let matrix = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	assert_eq!(matrix * 2, Matrix::rows([[2, 8, 10], [4, 6, 12]]));
}

#[test]
fn matrix_mul_assign_scalar() {
	let mut matrix = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	matrix *= 2;
	
	assert_eq!(matrix, Matrix::rows([[2, 8, 10], [4, 6, 12]]));
}

#[test]
fn matrix_div_scalar() {
	let matrix = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	assert_eq!(matrix / 2, Matrix::rows([[0, 2, 2], [1, 1, 3]]));
}

#[test]
fn matrix_div_assign_scalar() {
	let mut matrix = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	matrix /= 2;
	
	assert_eq!(matrix, Matrix::rows([[0, 2, 2], [1, 1, 3]]));
}

#[test]
fn matrix_add() {
	let a = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	let b = Matrix::rows([[0, 1, 1], [2, 3, 5]]);
	
	assert_eq!(a + b, Matrix::rows([[1, 5, 6], [4, 6, 11]]));
}

#[test]
fn matrix_add_assign() {
	let mut a = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	let b = Matrix::rows([[0, 1, 1], [2, 3, 5]]);
	
	a += b;
	
	assert_eq!(a, Matrix::rows([[1, 5, 6], [4, 6, 11]]));
}

#[test]
fn matrix_sub() {
	let a = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	let b = Matrix::rows([[0, 1, 1], [2, 3, 5]]);
	
	assert_eq!(a - b, Matrix::rows([[1, 3, 4], [0, 0, 1]]));
}

#[test]
fn matrix_sub_assign() {
	let mut a = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	let b = Matrix::rows([[0, 1, 1], [2, 3, 5]]);
	
	a -= b;
	
	assert_eq!(a, Matrix::rows([[1, 3, 4], [0, 0, 1]]));
}

#[test]
fn matrix_neg() {
	let matrix = Matrix::rows([[1, -4, 5], [-2, 3, -6]]);
	
	assert_eq!(-matrix, Matrix::rows([[-1, 4, -5], [2, -3, 6]]));
}

#[test]
fn matrix_mul() {
	let a = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	let b = Matrix::rows([[0, 1], [1, 2], [3, 5]]);
	
	assert_eq!(a * b, Matrix::rows([[19, 34], [21, 38]]));
}

#[test]
fn matrix_mul_assign() {
	let mut a = Matrix::rows([[0, 1], [1, 2], [3, 5]]);
	
	let b = Matrix::rows([[2, 3], [3, 5]]);
	
	a *= b;
	
	assert_eq!(a, Matrix::rows([[3, 5], [8, 13], [21, 34]]));
}

#[test]
fn matrix_mul_add() {
	let a = Matrix::rows([[1, 4, 5], [2, 3, 6]]);
	
	let b = Matrix::rows([[0, 1], [1, 2], [3, 5]]);
	
	let c = Matrix::rows([[2, 3], [5, 7]]);
	
	assert_eq!(a.mul_add(b, c), Matrix::rows([[21, 37], [26, 45]]));
}

#[test]
fn matrix_mul_add_assign() {
	let mut a = Matrix::rows([[0, 1], [1, 2], [3, 5]]);
	
	let b = Matrix::rows([[2, 3], [3, 5]]);
	
	let c = Matrix::rows([[2, 3], [5, 7], [11, 13]]);
	
	a.mul_add_assign(b, c);
	
	assert_eq!(a, Matrix::rows([[5, 8], [13, 20], [32, 47]]));
}

#[test]
fn matrix_div() {
	let a = Matrix::rows([[3., 5.], [8., 13.], [21., 34.]]);
	
	let b = Matrix::rows([[2., 3.], [3., 5.]]);
	
	assert_eq!(a / b, Matrix::rows([[0., 1.], [1., 2.], [3., 5.]]));
}

#[test]
fn matrix_div_left() {
	let a = Matrix::cols([[3., 5.], [8., 13.], [21., 34.]]);
	
	let b = Matrix::rows([[2., 3.], [3., 5.]]);
	
	assert_eq!(a.div_left(b), Matrix::cols([[0., 1.], [1., 2.], [3., 5.]]));
}

#[test]
fn matrix_div_assign() {
	let mut a = Matrix::rows([[3., 5.], [8., 13.], [21., 34.]]);
	
	let b = Matrix::rows([[2., 3.], [3., 5.]]);
	
	a /= b;
	
	assert_eq!(a, Matrix::rows([[0., 1.], [1., 2.], [3., 5.]]));
}

#[test]
fn matrix_inv() {
	let matrix = Matrix::rows([[2., 3.], [3., 5.]]);
	
	assert_eq!(matrix.inv(), Matrix::rows([[5., -3.], [-3., 2.]]));
}

#[test]
fn matrix_pow() {
	let matrix = Matrix::rows([[2., 3.], [3., 5.]]);
	
	assert_eq!(matrix.pow(0), Matrix::rows([[1., 0.], [0., 1.]]));
	assert_eq!(matrix.pow(1), Matrix::rows([[2., 3.], [3., 5.]]));
	assert_eq!(matrix.pow(3), Matrix::rows([[89., 144.], [144., 233.]]));
	assert_eq!(matrix.pow(-1), Matrix::rows([[5., -3.], [-3., 2.]]));
	assert_eq!(matrix.pow(-3), Matrix::rows([[233., -144.], [-144., 89.]]));
}


tests/vector.rs
use sized_matrix::{Vector, Dot};
use higher_order_functions::{Init, Section};

#[test]
fn vector_index() {
	let vector = Vector::vector([5, 6, 7, 8]);
	
	for i in 0..4 {
		assert_eq!(vector[i], i + 5);
	}
}

#[test]
fn vector_index_mut() {
	let mut vector = Vector::vector([0, 0, 0, 0]);
	
	for i in 0..4 {
		vector[i] = i;
	}
	
	assert_eq!(vector, Vector::<_, 4>::init(|x: usize| x));
}

#[test]
fn init_vector() {
	let vector = Vector::<_, 4>::init(|x: usize| x);
	
	for i in 0..4 {
		assert_eq!(vector[i], i);
	}
}

#[test]
fn vector_vector() {
	let vector = Vector::vector([0, 1, 2, 3]);
	
	assert_eq!(vector, Vector::<_, 4>::init(|x: usize| x));
}

#[test]
fn vector_section() {
	let vector = Vector::vector([0, 1, 2, 3]);
	
	assert_eq!(vector.section(1), Vector::vector([1, 2]));
}

#[test]
fn vector_dot() {
	let a = Vector::vector([0, 1, 1, 3, 5]);
	
	let b = Vector::vector([2, 3, 5, 7, 11]);
	
	assert_eq!(a.dot(b), 84);
}


tests/version-numbers.rs
#[test]
fn test_readme_deps() {
	version_sync::assert_markdown_deps_updated!("README.md");
}

#[test]
fn test_changelog_deps() {
	version_sync::assert_markdown_deps_updated!("CHANGELOG.md");
}

#[test]
fn test_html_root_url() {
	version_sync::assert_html_root_url_updated!("src/lib.rs");
}


src/traits.rs
use num_traits::Num;

use core::mem::{MaybeUninit, transmute_copy, forget};

/// Types which are used to scale Matrices.
pub trait Scalar { }
impl<T: Num> Scalar for T { }

/// Unary operator for transposing a value.
pub trait Transpose {
	type Output;
	
	/// Returns the transpose of `self`.
	fn transpose(self) -> Self::Output;
}

impl<T, const M: usize, const N: usize> Transpose for [[T; M]; N] {
	type Output = [[T; N]; M];
	
	fn transpose(self) -> [[T; N]; M] {
		let consumed: [[MaybeUninit<T>; M]; N] = unsafe { transmute_copy(&self) };
		forget(self);
		let mut contents: [[MaybeUninit<T>; N]; M] = unsafe { MaybeUninit::uninit().assume_init() };
		
		for i in 0..M {
			for j in 0..N {
				contents[i][j].write(unsafe { consumed[j][i].assume_init_read() });
			}
		}
		
		unsafe { transmute_copy(&contents) }
	}
}

/// Binary operator for the dot product operation.
pub trait Dot<TRhs = Self> {
	type Output;
	
	/// Returns `self · rhs`.
	fn dot(self, rhs: TRhs) -> Self::Output;
}


src/vector.rs
use super::{Matrix, Dot};

use higher_order_functions::{Init, Section};

use core::ops::{
	Mul,
	Index, IndexMut,
};

use num_traits::{MulAdd, Zero};

/// A special case type alias for Matrices with a single column.
pub type Vector<T, const M: usize> = Matrix<T, M, 1>;

impl<T, const M: usize> Vector<T, M> {
	/// Construct a column vector from an array.
	///
	/// # Examples
	///
	/// Constructing a `Vector`:
	///
	/// ```rust
	/// use sized_matrix::{Vector, Matrix};
	///
	/// let vector: Vector<i32, 3> = Vector::vector([1, 2, 3]);
	///
	/// assert_eq!(vector, Matrix::cols([[1, 2, 3]]));
	/// ```
	pub fn vector(contents: [T; M]) -> Self {
		Self::cols([contents])
	}
}

impl<T, const M: usize> Init<T, usize> for Vector<T, M> {
	fn init_with<F: FnMut(usize) -> T>(_: (), mut elem: F) -> Self {
		Self::init(|[row, _]: [usize; 2]| elem(row))
	}
}

impl<T, const M: usize> Index<usize> for Vector<T, M> {
	type Output = T;
	
	fn index(&self, index: usize) -> &T {
		&self[[index, 0]]
	}
}

impl<T, const M: usize> IndexMut<usize> for Vector<T, M> {
	fn index_mut(&mut self, index: usize) -> &mut T {
		&mut self[[index, 0]]
	}
}

impl<T: Copy, const M: usize, const M_OUT: usize> Section<usize, Vector<T, M_OUT>> for Vector<T, M> {
	fn section(&self, offset: usize) -> Vector<T, M_OUT> {
		self.section([offset, 0])
	}
}

impl<TLhs: Copy, TRhs: Copy, TOutput, const M: usize> Dot<Vector<TRhs, M>> for Vector<TLhs, M> where
	TLhs: Mul<TRhs, Output = TOutput> + MulAdd<TRhs, TOutput, Output = TOutput>,
	TOutput: Zero,
{
	type Output = TOutput;
	
	fn dot(self, rhs: Vector<TRhs, M>) -> Self::Output {
		let mut result = Self::Output::zero();
		for k in 0..M {
			result = self[k].mul_add(rhs[k], result);
		}
		result
	}
}


src/array_helpers.rs
use super::Transpose;

use core::mem::{MaybeUninit, transmute_copy, forget, replace};

// FIXME: Replace with some other library that does this once they start supporting const generics

pub trait MapArray<T, const N: usize> {
	fn map<U, F: FnMut(T) -> U>(self, f: F) -> [U; N];
}

impl<T, const N: usize> MapArray<T, N> for [T; N] {
	fn map<U, F: FnMut(T) -> U>(self, mut elem: F) -> [U; N] {
		let mut consumed: [MaybeUninit<T>; N] = unsafe { transmute_copy(&self) };
		forget(self);
		let mut contents: [MaybeUninit<U>; N] = MaybeUninit::uninit_array();
		
		for i in 0..N {
			contents[i] = MaybeUninit::new(elem(unsafe { replace(&mut consumed[i], MaybeUninit::uninit()).assume_init() }));
		}
		
		let res: [U; N] = unsafe { transmute_copy(&contents) };
		forget(contents);
		res
	}
}

impl<T, const M: usize, const N: usize> Transpose for [[T; M]; N] {
	type Output = [[T; N]; M];
	
	fn transpose(self) -> [[T; N]; M] {
		let mut consumed: [[MaybeUninit<T>; M]; N] = unsafe { transmute_copy(&self) };
		forget(self);
		let mut contents: [[MaybeUninit<T>; N]; M] = unsafe { MaybeUninit::uninit().assume_init() };
		
		for i in 0..M {
			for j in 0..N {
				contents[i][j] = replace(&mut consumed[j][i], MaybeUninit::uninit());
			}
		}
		
		let res: [[T; N]; M] = unsafe { transmute_copy(&contents) };
		forget(contents);
		res
	}
}


src/lib.rs
//! Sized matrices using const generics for better type checking and performance.
//!
//! ```rust
//! use sized_matrix::{Matrix, Vector};
//!
//! let a: Matrix<i32, 3, 4> = Matrix::rows([
//!     [ 1,  2,  3,  4],
//!     [ 5,  6,  7,  8],
//!     [ 9, 10, 11, 12],
//! ]);
//!
//! let b: Matrix<i32, 4, 2> = Matrix::rows([
//!     [ 0,  1],
//!     [ 1,  2],
//!     [ 3,  5],
//!     [ 8, 13],
//! ]);
//!
//! let c: Matrix<i32, 3, 2> = a * b;
//!
//! assert_eq!(c, Matrix::rows([
//!     [ 43,  72],
//!     [ 91, 156],
//!     [139, 240],
//! ]));
//!
//! let d: Vector<i32, 2> = Matrix::vector([-1, 1]);
//!
//! let e: Vector<i32, 3> = c * d;
//!
//! assert_eq!(e, Matrix::vector([
//!      29,
//!      65,
//!     101,
//! ]));
//! ```
//!
//! To use this, add it as a dependency to your Cargo.toml:
//! ```toml
//! [dependencies]
//! sized_matrix = "^0.3.0"
//! ```

#![no_std]

#![feature(generic_associated_types)]
#![feature(maybe_uninit_uninit_array)]
#![feature(negative_impls)]

#![doc(html_root_url = "https://docs.rs/sized_matrix/0.3.0")]

mod traits;
pub use traits::*;

mod matrix;
pub use matrix::*;

mod vector;
pub use vector::*;

// Include the readme and changelog as hidden documentation so they're tested by cargo test
#[doc = include_str!("../README.md")]
#[doc = include_str!("../CHANGELOG.md")]
type _Doctest = ();
